### PR TITLE
Bugfix/2601 parser fail

### PR DESCRIPTION
**Pull request by mitzimorris**
#### Submission Checklist

- [x] Run unit tests: `./runTests.py src/test/unit`
- [x] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary
Fix for https://github.com/stan-dev/stan/issues/2607

#### Intended Effect
Brings parser into line with language per reference manual:
```
Variable names will not conflict with the following block identifiers,

functions, model, data, parameters, quantities,
transformed, generated
```

#### How to Verify
Unit test `src/test/unit/lang/parser/term_grammar_test.cpp` tests that model `src/test/test-models/good/parser-generator/allowed_var_names.stan` is parseable.

#### Side Effects
Parser no longer has informative error message when a program block is missing a close brace and is followed by another program block.  Test model `src/test/test-models/bad/expect_statement_seq_close_brace_3.stan`:

```
transformed data {
  int vs[2,3];
  int z;
  for (v in vs[1]) {
    z = 3;
  }
  
parameters {
  real y;
}
model {
  y ~ normal(0,1);
}
```
fails to parse with confusing error message:

```
SYNTAX ERROR, MESSAGE(S) FROM PARSER:
Variable "parameters" does not exist.
 error in 'stan/src/test/test-models/bad/expect_statement_seq_close_brace_3.stan' at line 8, column 11
  -------------------------------------------------
     6:   }
     7:   
     8: parameters {
                  ^
     9:   real y;
  -------------------------------------------------

```

#### Documentation
NA

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):  Columbia University

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by mitzimorris** [comment]
@VMatthijs this is a pretty simple change - removed some logic causing the parser to fail on variables named `data`, also `parameters`, `tranformed` and `generated`.

**Comment by VMatthijs** [comment]
Will do! :-)

File: src/test/test-models/good/parser-generator/allowed_var_names.stan [hunk 1]
```diff
@@ -0,0 +1,8 @@
+functions {
+  real foo1(vector data) {
+    return data[1];
+  }
+  real foo2(real data) {
+    return data;
+  }
+}
```
[the remark this answers is not in the document]
Comment (mitzimorris) [reply src/test/test-models/good/parser-generator/allowed_var_names.stan]: I'll update test model to test all block keywords in all contexts.

this restriction was added to catch errors arising from forgetting to close off one program block before starting the next one, but this implementation is wrong and contradicts the reference manual, (the de-facto language spec).  backing this out will result in a misleading error for this particular situation. not good, but better than current behavoir - at least that's my call here.

[the remark this answers is not in the document]
Comment (VMatthijs) [reply src/test/test-models/good/parser-generator/allowed_var_names.stan]: Great! Once you add them, I'll approve.

[the remark this answers is not in the document]
Comment (VMatthijs) [reply src/test/test-models/good/parser-generator/allowed_var_names.stan]: @mitzimorris , I don't want to hold up this PR by demanding lots of tests. I will leave it to your judgement whether to add the extra ones or not.

**Comment by mitzimorris** [comment]
@VMatthijs thanks for asking for more tests - attempts to declare block variables with reserved names fails, e.g.:

```
data {
  int transformed;
  int parameters;
  int generated;
  int model;
}
```
fails with message:
```
Variable identifier (name) may not be reserved word
    found identifier=transformed
 error in 'src/test/test-models/good/parser-generator/allowed_var_names.stan' at line 28, column 6
  -------------------------------------------------
    26: }
    27: data {
    28:   int transformed;
             ^
    29:   int parameters;
  -------------------------------------------------

PARSER EXPECTED: <identifier>
" thrown in the test body.
```

**Comment by mitzimorris** [comment]
My reading of the Stan Language Reference manual was wrong; what the statement:

```
Variable names will not conflict with the following block identifiers,

functions, model, data, parameters, quantities,
transformed, generated
```

is saying is the words in the list are reserved words in the Stan language and cannot be used as variable names. 

Closing this PR because it's not a bug, it's a feature - the parser is doing the right thing here, although the error message is confusing.

**Comment by VMatthijs** [comment]
Great! I agree that the wording in the manual is confusing though. I interpreted it the same way as you did, though I wasn't quite sure.

**Pull request closed without merging**